File: src/main/java/model/Candidato.java
package model;

import kontroller.Usuario;
import org.hibernate.annotations.GenericGenerator;
import org.hibernate.annotations.LazyCollection;
import org.hibernate.annotations.LazyCollectionOption;

import javax.persistence.*;
import java.util.List;
import java.util.Objects;
import java.util.Set;

@Entity
public class Candidato {

	@Id
	@GeneratedValue(strategy = GenerationType.AUTO, generator = "native")
	@GenericGenerator(name = "native", strategy = "native")
	private Long id;

	private String nome;

	@OneToOne(fetch = FetchType.LAZY, cascade = CascadeType.PERSIST)
	private Usuario usuario;

	@ManyToMany
	@LazyCollection(LazyCollectionOption.FALSE)
	private List<Competencia> competencias;

	@OneToMany(fetch = FetchType.EAGER)
	private Set<Avaliacao> avaliacoes;

	public Candidato() {
	}

	public Candidato(String nome) {
		this.nome = nome;
	}

	public Candidato(Long id, String nome) {
		this.nome = nome;
		this.id = id;
	}


	public Candidato(Long id,String nome, Set<Avaliacao> avaliacoes) {
		this.id = id;
		this.nome = nome;
		this.avaliacoes = avaliacoes;
	}

	public Long getId() {
		return id;
	}

	public void setId(Long id) {
		this.id = id;
	}

	public List<Competencia> getCompetencias() {
		return competencias;
	}

	public void setCompetencias(List<Competencia> competencias) {
		this.competencias = competencias;
	}

	public Set<Avaliacao> getAvaliacoes() {
		return avaliacoes;
	}

	public void setAvaliacoes(Set<Avaliacao> avaliacoes) {
		this.avaliacoes = avaliacoes;
	}

	public String getNome() {
		return nome;
	}

	public void setNome(String nome) {
		this.nome = nome;
	}

	public Usuario getUsuario() {
		return usuario;
	}

	public void setUsuario(Usuario usuario) {
		this.usuario = usuario;
	}

	@Override
	public boolean equals(Object o) {
		if (this == o) return true;
		if (o == null || getClass() != o.getClass()) return false;
		Candidato candidato = (Candidato) o;
		return Objects.equals(id, candidato.id) &&
				Objects.equals(nome, candidato.nome);
	}

	@Override
	public int hashCode() {

		return Objects.hash(id, nome);
	}
}

File: src/main/java/model/Candidato.kt
package model

import org.hibernate.annotations.GenericGenerator
import org.hibernate.annotations.LazyCollection
import org.hibernate.annotations.LazyCollectionOption
import java.io.Serializable
import javax.persistence.*

@Entity
class Candidato : Serializable {

    @Id
    @GeneratedValue(strategy = GenerationType.AUTO, generator = "native")
    @GenericGenerator(name = "native", strategy = "native")
    var id: Long = 0L

    var nome: String = ""

    @OneToOne(fetch = FetchType.LAZY, cascade = [CascadeType.PERSIST])
    var usuario: Usuario? = null

    @ManyToMany
    @LazyCollection(LazyCollectionOption.FALSE)
    var competencias: List<Competencia> = ArrayList()

    @OneToMany(fetch = FetchType.EAGER)
    var avaliacoes: Set<Avaliacao> = emptySet()

    constructor()

    constructor(nome: String) {
        this.nome = nome
    }

    constructor(id: Long, nome: String) {
        this.nome = nome
        this.id = id
    }

    constructor(id: Long, nome: String, avaliacoes: Set<Avaliacao>) {
        this.id = id
        this.nome = nome
        this.avaliacoes = avaliacoes
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Candidato

        if (id != other.id) return false
        if (nome != other.nome) return false

        return true
    }

    override fun hashCode(): Int {
        var result = id.hashCode()
        result = 31 * result + nome.hashCode()
        return result
    }


}